src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java

```diff
@@ -0,0 +1,570 @@
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getPasswordlessDevicesTable()
+                + " WHERE device_id_hash = ?";
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, deviceIdHash);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void deleteDevicesByPhoneNumber_Transaction(Start start, Connection con, @Nonnull String phoneNumber)
+            throws SQLException {
+
+        String QUERY = "SELECT device_id_hash FROM " + Config.getConfig(start).getPasswordlessDevicesTable()
+                + " WHERE phone_number = ?";
+
+        List<String> deviceIdHashes = new ArrayList<>();
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, phoneNumber);
+            ResultSet result = pst.executeQuery();
+            while (result.next()) {
+                deviceIdHashes.add(result.getString("device_id_hash"));
+            }
+        }
+        deleteRowsByDeviceIdHashList_Transaction(con, Config.getConfig(start).getPasswordlessDevicesTable(),
+                deviceIdHashes);
+    }
+
+    public static void deleteDevicesByEmail_Transaction(Start start, Connection con, @Nonnull String email)
+            throws SQLException, StorageQueryException {
+
+        String QUERY = "SELECT device_id_hash FROM " + Config.getConfig(start).getPasswordlessDevicesTable()
+                + " WHERE email = ?";
+
+        List<String> deviceIdHashes = new ArrayList<>();
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, email);
+            ResultSet result = pst.executeQuery();
+            while (result.next()) {
+                deviceIdHashes.add(result.getString("device_id_hash"));
+            }
+        }
+        deleteRowsByDeviceIdHashList_Transaction(con, Config.getConfig(start).getPasswordlessDevicesTable(),
+                deviceIdHashes);
+    }
+
+    private static void createCode_Transaction(Start start, Connection con, PasswordlessCode code) throws SQLException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getPasswordlessCodesTable()
+                + "(code_id, device_id_hash, link_code_hash, created_at)" + " VALUES(?, ?, ?, ?)";
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, code.id);
+            pst.setString(2, code.deviceIdHash);
+            pst.setString(3, code.linkCodeHash);
+            pst.setLong(4, code.createdAt);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void createCode(Start start, PasswordlessCode code)
+            throws StorageTransactionLogicException, StorageQueryException {
+        start.startTransaction(con -> {
+            Connection sqlCon = (Connection) con.getConnection();
+
+            try {
+                if (PasswordlessQueries.getDevice_Transaction(start, sqlCon, code.deviceIdHash) == null) {
```

Comment:
This should be done by the foreign key constraint.

Reply:
noted.